pkg/stanza/docs/operators/journald_input.md

```diff
@@ -33,7 +34,23 @@ The `journald_input` operator will use the `__REALTIME_TIMESTAMP` field of the j
 - type: journald_input
   priority: emerg..err
 ```
-#### Simple journald input
+
+#### Matches
```

Comment:
Should there be a mechanism to prevent specifying both `units:` and `matches:` configuration for a single input?

Reply:
Not really sure about it, as you can do it in `journalctl` as well

Reply:
My testing with `journalctl` trying to use both `units` and `matches` syntax gave unexpected results - specifically NO LOG DATA.  At a minimum, there should be a warning in the docs that mixing `units` and `matches` could lead to unexpected results.

- One Unit: 
  - `journalctl --utc --output=short --follow --priority info --unit ssh.service`
  - Only SSH unit logs ... as expected
- One Match Rule (with 2 terms):
  - `journalctl --utc --output=short --follow --priority info _TRANSPORT=kernel PRIORITY=5`
  - Only kernel logs ... as expected
- Two Match Rules: 
  - `journalctl --utc --output=short --follow --priority info _SYSTEMD_UNIT=ssh.service + _TRANSPORT=kernel PRIORITY=5`
  - Both log streams present ... as expected
- Unit and Match: 
  - `journalctl --utc --output=short --follow --priority info --unit ssh.service _TRANSPORT=kernel PRIORITY=5`
  - NO LOGS
- Unit _plus_ Match:
  - `journalctl --utc --output=short --follow --priority info --unit ssh.service + _TRANSPORT=kernel PRIORITY=5`
  - Invalid syntax: `"+" can only be used between terms`
- Match and Unit:
  - `journalctl --utc --output=short --follow --priority info _TRANSPORT=kernel PRIORITY=5 --unit ssh.service`
  - NO LOGS
- Match _plus_ Unit:
  - `journalctl --utc --output=short --follow --priority info _TRANSPORT=kernel PRIORITY=5 + --unit ssh.service`
  - Invalid syntax: `"+" can only be used between terms`

Tested on Ubuntu 20.04 host with a noisy iptables LOG rule (kernel log stream) and inbound SSH probes.

Reply:
I wouldn't say the the behavior is not fully unexpected. Look at the following example:

```
➜  ~ journalctl --utc --output=json --no-tail --priority info --unit=user@1000.service --unit=init.scope _COMM=wireplumber + _COMM=systemd | jq '{unit: ._SYSTEMD_UNIT, comm: ._COMM} | tostring' | sort | uniq
"{\"unit\":\"init.scope\",\"comm\":\"systemd\"}"
"{\"unit\":\"user@1000.service\",\"comm\":\"systemd\"}"
"{\"unit\":\"user@1000.service\",\"comm\":\"wireplumber\"}
```

How I understand the behavior is:

1. take all entries which matches the units configuration
2. apply matches configuration to the entries which has been selected in `1`

Having both matches and units can be actually helpful, if you want to select multiple units with the same additional parameter, so you can do it in the following way:

```
units:
- a
- b
- c
- d
matches:
- _COMM=systemd
```

instead of:

```
matches:
- _SYSTEMD_UNIT: a
  _COMM: systemd
- _SYSTEMD_UNIT: b
  _COMM: systemd
- _SYSTEMD_UNIT: c
  _COMM: systemd
- _SYSTEMD_UNIT: d
  _COMM: systemd
```

Anyway, I would allow either `matches` or `units` to avoid confusion and non-well documented behavior. It will be also easier to deal with it if we decide to go with jorunald gateway path

Reply:
> My testing with journalctl trying to use both units and matches syntax gave unexpected results - specifically NO LOG DATA.

I don't agree that the results are unexpected - in fact, they are exactly what I would expect. The `units` and `matches` conditions are AND-ed, same as `priority` and `units` are AND-ed in the current version of the receiver.

> Anyway, I would allow either `matches` or `units` to avoid confusion and non-well documented behavior. It will be also easier to deal with it if we decide to go with jorunald gateway path

I don't agree with this. Both `matches` and `units` should be possible to set together, as this makes perfect sense - as you have described in your comment @sumo-drosiek, and as it is possible in `journalctl`. The argument about journald gateway doesn't make much sense, as the `matches` configuration as implemented is already incompatible with the journald gateway (which apparently only accepts a flat list of AND-ed conditions). We are already not compatible with the journald gateway with the `matches` and I think it makes most sense to allow the user to do the same thing as they can do with `journalctl`.

Reply:
> The argument about journald gateway doesn't make much sense, as the matches configuration as implemented is already incompatible with the journald gateway (which apparently only accepts a flat list of AND-ed conditions).

We are compatible with `journald-gateway` and will be anyway, as we are able to get all entries based on the configuration (there is no requirement to do it in one query). Only thing is that allowing to have both `matches` and `units` introduces some logic complexity which could be avoided.